File: pkg/dart2cpp/demo/comprehensive_demo.dart
/// 综合测试演示 - 包含所有Dart语法特性
/// 
/// 这是一个完整的测试文件，包含了所有主要的Dart语法特性
/// 用于测试dart2cpp转换器的完整性和正确性

import 'dart:async';
import 'dart:math' as math;

void main() async {
  print('🚀 Dart2CPP 综合测试开始');
  print('=' * 60);

  // 1. 基础语法测试
  testBasicSyntax();
  
  // 2. 面向对象测试
  testObjectOriented();
  
  // 3. 集合操作测试
  testCollections();
  
  // 4. 异步编程测试
  await testAsyncProgramming();
  
  // 5. 泛型测试
  testGenerics();
  
  // 6. 异常处理测试
  testExceptionHandling();
  
  // 7. 高级特性测试
  testAdvancedFeatures();

  print('=' * 60);
  print('✅ 所有测试完成');
}

/// 1. 测试基础语法
void testBasicSyntax() {
  print('\n📌 测试基础语法');

  // 基本数据类型
  int intValue = 42;
  double doubleValue = 3.14159;
  bool boolValue = true;
  String stringValue = 'Hello, Dart2CPP!';
  
  print('  基本类型: $intValue, $doubleValue, $boolValue, $stringValue');

  // 运算符测试
  int a = 10, b = 3;
  print('  算术运算: ${a + b}, ${a - b}, ${a * b}, ${a / b}, ${a % b}, ${a ~/ b}');
  print('  比较运算: ${a == b}, ${a != b}, ${a > b}, ${a < b}');
  print('  逻辑运算: ${boolValue && false}, ${boolValue || false}, ${!boolValue}');

  // 控制流测试
  int sum = 0;
  for (int i = 1; i <= 5; i++) {
    sum += i;
  }
  print('  for循环求和: $sum');

  // 条件语句
  String grade = intValue >= 90 ? 'A' : intValue >= 80 ? 'B' : 'C';
  print('  条件运算: $grade');

  // switch语句
  switch (grade) {
    case 'A':
      print('  优秀');
      break;
    case 'B':
      print('  良好');
      break;
    default:
      print('  需要努力');
  }

  // 函数调用
  int result = addNumbers(5, 3);
  print('  函数调用结果: $result');

  // 可选参数
  greetPerson('Alice');
  greetPerson('Bob', 'Mr.');

  print('  ✓ 基础语法测试完成');
}

/// 2. 测试面向对象
void testObjectOriented() {
  print('\n📌 测试面向对象');

  // 基本类
  var person = Person('Alice', 25);
  person.introduce();
  person.celebrateBirthday();

  // 继承
  var student = Student('Bob', 20, 'S001');
  student.introduce();
  student.study();

  // 接口实现
  var car = Car();
  car.start();
  car.stop();

  // 抽象类
  var circle = Circle(5.0);
  print('  圆形面积: ${circle.area()}');

  var rectangle = Rectangle(4.0, 6.0);
  print('  矩形面积: ${rectangle.area()}');

  // 枚举
  var status = Status.active;
  print('  状态: ${status.name}, 索引: ${status.index}');

  // 混入
  var musician = Musician('Charlie');
  musician.perform();

  print('  ✓ 面向对象测试完成');
}

/// 3. 测试集合操作
void testCollections() {
  print('\n📌 测试集合操作');

  // List 操作
  List<int> numbers = [1, 2, 3, 4, 5];
  print('  原始列表: $numbers');
  
  numbers.add(6);
  print('  添加元素后: $numbers');
  
  var doubled = numbers.map((n) => n * 2).toList();
  print('  映射操作: $doubled');
  
  var evens = numbers.where((n) => n % 2 == 0).toList();
  print('  过滤偶数: $evens');

  // Set 操作
  Set<String> fruits = {'apple', 'banana', 'orange'};
  fruits.add('apple'); // 重复元素会被忽略
  print('  水果集合: $fruits');

  // Map 操作
  Map<String, int> scores = {'Alice': 95, 'Bob': 87, 'Charlie': 92};
  print('  分数映射: $scores');
  print('  Alice的分数: ${scores['Alice']}');

  // 集合方法
  int sum = numbers.reduce((a, b) => a + b);
  print('  列表求和: $sum');
  
  bool hasEven = numbers.any((n) => n % 2 == 0);
  print('  包含偶数: $hasEven');

  print('  ✓ 集合操作测试完成');
}

/// 4. 测试异步编程
Future<void> testAsyncProgramming() async {
  print('\n📌 测试异步编程');

  // Future 基础
  String result = await Future.delayed(
    Duration(milliseconds: 10), 
    () => 'Async Result'
  );
  print('  异步结果: $result');

  // Future 组合
  List<String> results = await Future.wait([
    Future.delayed(Duration(milliseconds: 5), () => 'A'),
    Future.delayed(Duration(milliseconds: 10), () => 'B'),
    Future.delayed(Duration(milliseconds: 15), () => 'C'),
  ]);
  print('  并行结果: $results');

  // Stream 测试
  print('  Stream测试:');
  await for (int value in generateNumbers(3)) {
    print('    生成数字: $value');
  }

  // 错误处理
  try {
    await riskyAsyncOperation();
  } catch (e) {
    print('  捕获异步异常: $e');
  }

  print('  ✓ 异步编程测试完成');
}

/// 5. 测试泛型
void testGenerics() {
  print('\n📌 测试泛型');

  // 泛型类
  var intBox = Box<int>(42);
  var stringBox = Box<String>('Hello');
  
  print('  整数盒子: ${intBox.getValue()}');
  print('  字符串盒子: ${stringBox.getValue()}');

  // 泛型方法
  print('  泛型方法: ${identity<int>(100)}');
  print('  泛型方法: ${identity<String>("test")}');

  // 多泛型参数
  var pair = Pair<String, int>('Alice', 25);
  print('  键值对: ${pair.first} -> ${pair.second}');

  // 类型约束
  var calculator = Calculator<int>();
  print('  泛型计算: ${calculator.add(5, 3)}');

  print('  ✓ 泛型测试完成');
}

/// 6. 测试异常处理
void testExceptionHandling() {
  print('\n📌 测试异常处理');

  // 基本异常捕获
  try {
    throw Exception('测试异常');
  } catch (e) {
    print('  捕获异常: $e');
  }

  // 特定异常类型
  try {
    validateAge(-5);
  } on ArgumentError catch (e) {
    print('  参数错误: ${e.message}');
  }

  // 自定义异常
  try {
    throw CustomException('自定义错误', 404);
  } on CustomException catch (e) {
    print('  自定义异常: ${e.message}, 代码: ${e.code}');
  }

  // finally 块
  try {
    performRiskyOperation();
  } catch (e) {
    print('  操作异常: $e');
  } finally {
    print('  清理资源');
  }

  print('  ✓ 异常处理测试完成');
}

/// 7. 测试高级特性
void testAdvancedFeatures() {
  print('\n📌 测试高级特性');

  // 闭包
  var counter = createCounter();
  print('  闭包计数: ${counter()}, ${counter()}, ${counter()}');

  // 高阶函数
  var numbers = [1, 2, 3, 4, 5];
  var processed = processNumbers(numbers, (n) => n * n);
  print('  高阶函数: $processed');

  // 扩展方法（如果支持）
  String text = 'hello world';
  print('  字符串处理: ${text.toUpperCase()}');

  // 空安全
  String? nullable = null;
  String safe = nullable ?? '默认值';
  print('  空安全: $safe');

  // 级联操作
  var list = <int>[]
    ..add(1)
    ..add(2)
    ..add(3);
  print('  级联操作: $list');

  print('  ✓ 高级特性测试完成');
}

// ============================================================================
// 辅助函数和类定义
// ============================================================================

/// 简单加法函数
int addNumbers(int a, int b) {
  return a + b;
}

/// 带可选参数的问候函数
void greetPerson(String name, [String? title]) {
  if (title != null) {
    print('  问候: Hello, $title $name');
  } else {
    print('  问候: Hello, $name');
  }
}

/// 基本人员类
class Person {
  String name;
  int age;
  
  Person(this.name, this.age);
  
  void introduce() {
    print('  我是$name，今年$age岁');
  }
  
  void celebrateBirthday() {
    age++;
    print('  🎉 生日快乐！现在$age岁了');
  }
}

/// 学生类（继承示例）
class Student extends Person {
  String studentId;
  
  Student(String name, int age, this.studentId) : super(name, age);
  
  void study() {
    print('  学生$name正在学习，学号：$studentId');
  }
}

/// 可驾驶接口
abstract class Drivable {
  void start();
  void stop();
}

/// 汽车类（接口实现）
class Car implements Drivable {
  bool isRunning = false;
  
  @override
  void start() {
    isRunning = true;
    print('  汽车启动');
  }
  
  @override
  void stop() {
    isRunning = false;
    print('  汽车停止');
  }
}

/// 抽象形状类
abstract class Shape {
  double area();
}

/// 圆形类
class Circle extends Shape {
  double radius;
  
  Circle(this.radius);
  
  @override
  double area() {
    return math.pi * radius * radius;
  }
}

/// 矩形类
class Rectangle extends Shape {
  double width;
  double height;
  
  Rectangle(this.width, this.height);
  
  @override
  double area() {
    return width * height;
  }
}

/// 状态枚举
enum Status {
  inactive,
  active,
  pending,
  completed
}

/// 表演者混入
mixin Performer {
  void perform() {
    print('  正在表演...');
  }
}

/// 音乐家类（使用混入）
class Musician with Performer {
  String name;
  
  Musician(this.name);
}

/// 泛型盒子类
class Box<T> {
  T _value;
  
  Box(this._value);
  
  T getValue() => _value;
  void setValue(T value) => _value = value;
}

/// 泛型键值对类
class Pair<T, U> {
  T first;
  U second;
  
  Pair(this.first, this.second);
}

/// 泛型身份函数
T identity<T>(T value) {
  return value;
}

/// 带类型约束的计算器
class Calculator<T extends num> {
  T add(T a, T b) {
    return (a + b) as T;
  }
}

/// 自定义异常类
class CustomException implements Exception {
  final String message;
  final int code;
  
  CustomException(this.message, this.code);
  
  @override
  String toString() => 'CustomException: $message (code: $code)';
}

/// 验证年龄函数
void validateAge(int age) {
  if (age < 0) {
    throw ArgumentError('年龄不能为负数');
  }
}

/// 执行风险操作
void performRiskyOperation() {
  throw Exception('风险操作失败');
}

/// 创建计数器闭包
Function createCounter() {
  int count = 0;
  return () {
    count++;
    return count;
  };
}

/// 处理数字的高阶函数
List<int> processNumbers(List<int> numbers, int Function(int) processor) {
  return numbers.map(processor).toList();
}

/// 生成数字的Stream
Stream<int> generateNumbers(int count) async* {
  for (int i = 1; i <= count; i++) {
    await Future.delayed(Duration(milliseconds: 10));
    yield i;
  }
}

/// 风险异步操作
Future<void> riskyAsyncOperation() async {
  await Future.delayed(Duration(milliseconds: 5));
  throw Exception('异步操作失败');
}

File: pkg/dart2cpp/demo/cpp/comprehensive_demo.cpp
#include "../../cpp/core/object.h"
#include <iostream>

// 工具宏定义
#define dart_print(value) \
    do { \
        std::cout << (value).toString().getValue() << std::endl; \
    } while(0)

#define dart_int(value) Int(value)
#define dart_double(value) Double(value)
#define dart_bool(value) Bool(value)
#define dart_string(value) String(value)

// ============================================================================
// 类: Person
// ============================================================================

class Person {
public:
  String name;
  Int age;
  Person(String name, Int age) : name(name), age(age) {
  }
  
  void introduce() {
    {
      dart_print(dart_string("  我是") + this->name.toString() + dart_string("，今年") + this->age.toString() + dart_string("岁"));
}
  }
  
  void celebrateBirthday() {
    {
      this->age = (this->age + dart_int(1));
      dart_print(dart_string("  🎉 生日快乐！现在") + this->age.toString() + dart_string("岁了"));
}
  }
  
};

// ============================================================================
// 类: Student
// ============================================================================

class Student : public Person {
public:
  String studentId;
  Student(String name, Int age, String studentId) : studentId(studentId), Person(name, age) {
  }
  
  void study() {
    {
      dart_print(dart_string("  学生") + this->name.toString() + dart_string("正在学习，学号：") + this->studentId.toString());
}
  }
  
};

// ============================================================================
// 类: Drivable
// ============================================================================

DART_INTERFACE(Drivable)
  DART_ABSTRACT_METHOD(void, start, ())
  DART_ABSTRACT_METHOD(void, stop, ())
DART_INTERFACE_END

// ============================================================================
// 类: Car
// ============================================================================

class Car : DART_IMPLEMENTS(Drivable) {
public:
  Bool isRunning = dart_bool(false);
  Car() {
  }
  
  void start() {
    {
      this->isRunning = dart_bool(true);
      dart_print(dart_string("  汽车启动"));
}
  }
  
  void stop() {
    {
      this->isRunning = dart_bool(false);
      dart_print(dart_string("  汽车停止"));
}
  }
  
};

// ============================================================================
// 类: Shape
// ============================================================================

DART_INTERFACE(Shape)
  DART_ABSTRACT_METHOD(Double, area, ())
DART_INTERFACE_END

// ============================================================================
// 类: Circle
// ============================================================================

class Circle : public Shape {
public:
  Double radius;
  Circle(Double radius) : radius(radius) {
  }
  
  Double area() {
    {
      return ((dart_double(3.141592653589793) * this->radius) * this->radius);
}
  }
  
};

// ============================================================================
// 类: Rectangle
// ============================================================================

class Rectangle : public Shape {
public:
  Double width;
  Double height;
  Rectangle(Double width, Double height) : width(width), height(height) {
  }
  
  Double area() {
    {
      return (this->width * this->height);
}
  }
  
};

// ============================================================================
// 类: Status
// ============================================================================

class Status : public _Enum {
public:
  Status inactive = ObjectPtr<Status>::createConst();
  Status active = ObjectPtr<Status>::createConst();
  Status pending = ObjectPtr<Status>::createConst();
  Status completed = ObjectPtr<Status>::createConst();
  List<Status> values = List<Status>::createConst({ObjectPtr<Status>::createConst(), ObjectPtr<Status>::createConst(), ObjectPtr<Status>::createConst(), ObjectPtr<Status>::createConst()});
  Status(Int #index, String #name) : _Enum(#index, #name) {
  }
  
  String _enumToString() {
    return dart_string("Status.") + this->_name.toString();
  }
  
};

// ============================================================================
// 类: Performer
// ============================================================================

DART_INTERFACE(Performer)
DART_INTERFACE_END

// ============================================================================
// 类: Musician
// ============================================================================

class Musician : public _Musician&Object&Performer {
public:
  String name;
  Musician(String name) : name(name) {
  }
  
};

// ============================================================================
// 类: Box
// ============================================================================

class Box {
private:
  Any _value;
public:
  Box(Any _value) : _value(_value) {
  }
  
  Any getValue() {
    return this->_value;
  }
  
  void setValue(Any value) {
    return this->_value = value;
  }
  
};

// ============================================================================
// 类: Pair
// ============================================================================

class Pair {
public:
  Any first;
  Any second;
  Pair(Any first, Any second) : first(first), second(second) {
  }
  
};

// ============================================================================
// 类: Calculator
// ============================================================================

class Calculator {
public:
  Calculator() {
  }
  
  Any add(Any a, Any b) {
    {
      return dart_cast<Any>((a + b));
}
  }
  
};

// ============================================================================
// 类: CustomException
// ============================================================================

class CustomException : DART_IMPLEMENTS(Exception) {
public:
  String message;
  Int code;
  CustomException(String message, Int code) : message(message), code(code) {
  }
  
  String toString() {
    return dart_string("CustomException: ") + this->message.toString() + dart_string(" (code: ") + this->code.toString() + dart_string(")");
  }
  
};

// ============================================================================
// 类: _Musician&Object&Performer
// ============================================================================

DART_INTERFACE(_Musician&Object&Performer)
DART_INTERFACE_END

DART_ASYNC_FUNCTION(void, main, ()) {
    DART_ASYNC_BEGIN
  {
    dart_print(dart_string("🚀 Dart2CPP 综合测试开始"));
    dart_print((dart_string("=") * dart_int(60)));
    testBasicSyntax();
    testObjectOriented();
    testCollections();
    DART_AWAIT(testAsyncProgramming());
    testGenerics();
    testExceptionHandling();
    testAdvancedFeatures();
    dart_print((dart_string("=") * dart_int(60)));
    dart_print(dart_string("✅ 所有测试完成"));
}
    DART_ASYNC_END
}

void testBasicSyntax() {
  {
    dart_print(dart_string("
📌 测试基础语法"));
    auto intValue = dart_int(42);
    auto doubleValue = dart_double(3.14159);
    auto boolValue = dart_bool(true);
    auto stringValue = dart_string("Hello, Dart2CPP!");
    dart_print(dart_string("  基本类型: ") + intValue.toString() + dart_string(", ") + doubleValue.toString() + dart_string(", ") + boolValue.toString() + dart_string(", ") + stringValue.toString());
    auto a = dart_int(10);
    auto b = dart_int(3);
    dart_print(dart_string("  算术运算: ") + (a + b).toString() + dart_string(", ") + (a - b).toString() + dart_string(", ") + (a * b).toString() + dart_string(", ") + (a / b).toString() + dart_string(", ") + (a % b).toString() + dart_string(", ") + a->~/(b).toString());
    dart_print(dart_string("  比较运算: ") + (a == b).toString() + dart_string(", ") + !((a == b)).toString() + dart_string(", ") + (a > b).toString() + dart_string(", ") + (a < b).toString());
    dart_print(dart_string("  逻辑运算: ") + boolValue && dart_bool(false).toString() + dart_string(", ") + boolValue || dart_bool(false).toString() + dart_string(", ") + !(boolValue).toString());
    auto sum = dart_int(0);
    for (auto i = dart_int(1);; (i <= dart_int(5)); i = (i + dart_int(1))) {
    sum = (sum + i);
}
    dart_print(dart_string("  for循环求和: ") + sum.toString());
    auto grade = (intValue >= dart_int(90)) ? dart_string("A") : (intValue >= dart_int(80)) ? dart_string("B") : dart_string("C");
    dart_print(dart_string("  条件运算: ") + grade.toString());
    label_40103: switch (grade) {
  case dart_string("A"):
    {
    dart_print(dart_string("  优秀"));
    break;
}
  case dart_string("B"):
    {
    dart_print(dart_string("  良好"));
    break;
}
  default:
    {
    dart_print(dart_string("  需要努力"));
}
    break;
}
    auto result = addNumbers(dart_int(5), dart_int(3));
    dart_print(dart_string("  函数调用结果: ") + result.toString());
    greetPerson(dart_string("Alice"));
    greetPerson(dart_string("Bob"), dart_string("Mr."));
    dart_print(dart_string("  ✓ 基础语法测试完成"));
}
}

void testObjectOriented() {
  {
    dart_print(dart_string("
📌 测试面向对象"));
    auto person = ObjectPtr<Person>(new Person(dart_string("Alice"), dart_int(25)));
    person->introduce();
    person->celebrateBirthday();
    auto student = ObjectPtr<Student>(new Student(dart_string("Bob"), dart_int(20), dart_string("S001")));
    student->introduce();
    student->study();
    auto car = ObjectPtr<Car>(new Car());
    car->start();
    car->stop();
    auto circle = ObjectPtr<Circle>(new Circle(dart_double(5.0)));
    dart_print(dart_string("  圆形面积: ") + circle->area().toString());
    auto rectangle = ObjectPtr<Rectangle>(new Rectangle(dart_double(4.0), dart_double(6.0)));
    dart_print(dart_string("  矩形面积: ") + rectangle->area().toString());
    auto status = ObjectPtr<Status>::createConst();
    dart_print(dart_string("  状态: ") + EnumName|get#name(status).toString() + dart_string(", 索引: ") + status->index.toString());
    auto musician = ObjectPtr<Musician>(new Musician(dart_string("Charlie")));
    musician->perform();
    dart_print(dart_string("  ✓ 面向对象测试完成"));
}
}

void testCollections() {
  {
    dart_print(dart_string("
📌 测试集合操作"));
    auto numbers = _GrowableList::_literal5(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
    dart_print(dart_string("  原始列表: ") + numbers.toString());
    numbers->add(dart_int(6));
    dart_print(dart_string("  添加元素后: ") + numbers.toString());
    auto doubled = numbers->map([&](Int n) { return (n * dart_int(2)); })->toList();
    dart_print(dart_string("  映射操作: ") + doubled.toString());
    auto evens = numbers->where([&](Int n) { return ((n % dart_int(2)) == dart_int(0)); })->toList();
    dart_print(dart_string("  过滤偶数: ") + evens.toString());
    auto fruits = ([&]() { const auto unnamed_var = ObjectPtr<_Set>(new _Set()); unnamed_var->add(dart_string("apple")); unnamed_var->add(dart_string("banana")); unnamed_var->add(dart_string("orange")); return unnamed_var; })();
    fruits->add(dart_string("apple"));
    dart_print(dart_string("  水果集合: ") + fruits.toString());
    auto scores = Map<String, Int>::createFromEntries({{dart_string("Alice"), dart_int(95)}, {dart_string("Bob"), dart_int(87)}, {dart_string("Charlie"), dart_int(92)}});
    dart_print(dart_string("  分数映射: ") + scores.toString());
    dart_print(dart_string("  Alice的分数: ") + scores->[](dart_string("Alice")).toString());
    auto sum = numbers->reduce([&](Int a, Int b) { return (a + b); });
    dart_print(dart_string("  列表求和: ") + sum.toString());
    auto hasEven = numbers->any([&](Int n) { return ((n % dart_int(2)) == dart_int(0)); });
    dart_print(dart_string("  包含偶数: ") + hasEven.toString());
    dart_print(dart_string("  ✓ 集合操作测试完成"));
}
}

DART_ASYNC_FUNCTION(Future<void>, testAsyncProgramming, ()) {
    DART_ASYNC_BEGIN
  {
    dart_print(dart_string("
📌 测试异步编程"));
    auto result = DART_AWAIT(Future::delayed(ObjectPtr<Duration>(new Duration(/*milliseconds:*/ dart_int(10))), [&]() { return dart_string("Async Result"); }));
    dart_print(dart_string("  异步结果: ") + result.toString());
    auto results = DART_AWAIT(Future::wait(_GrowableList::_literal3(Future::delayed(ObjectPtr<Duration>(new Duration(/*milliseconds:*/ dart_int(5))), [&]() { return dart_string("A"); }), Future::delayed(ObjectPtr<Duration>(new Duration(/*milliseconds:*/ dart_int(10))), [&]() { return dart_string("B"); }), Future::delayed(ObjectPtr<Duration>(new Duration(/*milliseconds:*/ dart_int(15))), [&]() { return dart_string("C"); }))));
    dart_print(dart_string("  并行结果: ") + results.toString());
    dart_print(dart_string("  Stream测试:"));
    {
    auto :stream = generateNumbers(dart_int(3));
    auto :for-iterator = ObjectPtr<_StreamIterator>(new _StreamIterator(:stream));
    try { /* try block */ } catch (const std::exception& e) { /* catch block */ }
}
    try {
    DART_AWAIT(riskyAsyncOperation());
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
    dart_print(dart_string("  ✓ 异步编程测试完成"));
}
    DART_ASYNC_END
}

void testGenerics() {
  {
    dart_print(dart_string("
📌 测试泛型"));
    auto intBox = ObjectPtr<Box>(new Box(dart_int(42)));
    auto stringBox = ObjectPtr<Box>(new Box(dart_string("Hello")));
    dart_print(dart_string("  整数盒子: ") + intBox->getValue().toString());
    dart_print(dart_string("  字符串盒子: ") + stringBox->getValue().toString());
    dart_print(dart_string("  泛型方法: ") + identity(dart_int(100)).toString());
    dart_print(dart_string("  泛型方法: ") + identity(dart_string("test")).toString());
    auto pair = ObjectPtr<Pair>(new Pair(dart_string("Alice"), dart_int(25)));
    dart_print(dart_string("  键值对: ") + pair->first.toString() + dart_string(" -> ") + pair->second.toString());
    auto calculator = ObjectPtr<Calculator>(new Calculator());
    dart_print(dart_string("  泛型计算: ") + calculator->add(dart_int(5), dart_int(3)).toString());
    dart_print(dart_string("  ✓ 泛型测试完成"));
}
}

void testExceptionHandling() {
  {
    dart_print(dart_string("
📌 测试异常处理"));
    try {
    throw DartException(Exception::(dart_string("测试异常")));
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
    try {
    validateAge((-dart_int(5)));
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
    try {
    throw DartException(ObjectPtr<CustomException>(new CustomException(dart_string("自定义错误"), dart_int(404))));
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
    try { /* try block */ } catch (const std::exception& e) { /* catch block */ }
    dart_print(dart_string("  ✓ 异常处理测试完成"));
}
}

void testAdvancedFeatures() {
  {
    dart_print(dart_string("
📌 测试高级特性"));
    auto counter = createCounter();
    dart_print(dart_string("  闭包计数: ") + counter().toString() + dart_string(", ") + counter().toString() + dart_string(", ") + counter().toString());
    auto numbers = _GrowableList::_literal5(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
    auto processed = processNumbers(numbers, [&](Int n) { return (n * n); });
    dart_print(dart_string("  高阶函数: ") + processed.toString());
    auto text = dart_string("hello world");
    dart_print(dart_string("  字符串处理: ") + text->toUpperCase().toString());
    auto nullable = nullptr;
    auto safe = ([&]() { String let_var = nullable; return (let_var == nullptr) ? dart_string("默认值") : let_var; })();
    dart_print(dart_string("  空安全: ") + safe.toString());
    auto list = ([&]() { List<Int> let_var = _GrowableList::(dart_int(0)); return ([&]() { let_var->add(dart_int(1)); let_var->add(dart_int(2)); let_var->add(dart_int(3)); return let_var; })(); })();
    dart_print(dart_string("  级联操作: ") + list.toString());
    dart_print(dart_string("  ✓ 高级特性测试完成"));
}
}

Int addNumbers(Int a, Int b) {
  {
    return (a + b);
}
}

void greetPerson(String name, String title) {
  {
    if (!((title == nullptr))) {
    dart_print(dart_string("  问候: Hello, ") + title.toString() + dart_string(" ") + name.toString());
} else {
    dart_print(dart_string("  问候: Hello, ") + name.toString());
}
}
}

Any identity(Any value) {
  {
    return value;
}
}

void validateAge(Int age) {
  {
    if ((age < dart_int(0))) {
    throw DartException(ObjectPtr<ArgumentError>(new ArgumentError(dart_string("年龄不能为负数"))));
}
}
}

void performRiskyOperation() {
  {
    throw DartException(Exception::(dart_string("风险操作失败")));
}
}

Function createCounter() {
  {
    auto count = dart_int(0);
    return [&]() { {
    count = (count + dart_int(1));
    return count;
} };
}
}

List<Int> processNumbers(List<Int> numbers, std::function<Int()> processor) {
  {
    return numbers->map(processor)->toList();
}
}

DART_ASYNC_FUNCTION(Stream<Int>, generateNumbers, (Int count)) {
    DART_ASYNC_BEGIN
  {
    for (auto i = dart_int(1);; (i <= count); i = (i + dart_int(1))) {
    DART_AWAIT(Future::delayed(ObjectPtr<Duration>(new Duration(/*milliseconds:*/ dart_int(10)))));
    co_yield i;  // C++20 coroutine
}
}
    DART_ASYNC_END
}

DART_ASYNC_FUNCTION(Future<void>, riskyAsyncOperation, ()) {
    DART_ASYNC_BEGIN
  {
    DART_AWAIT(Future::delayed(ObjectPtr<Duration>(new Duration(/*milliseconds:*/ dart_int(5)))));
    throw DartException(Exception::(dart_string("异步操作失败")));
}
    DART_ASYNC_END
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    {
      dart_print(dart_string("🚀 Dart2CPP 综合测试开始"));
      dart_print((dart_string("=") * dart_int(60)));
      testBasicSyntax();
      testObjectOriented();
      testCollections();
      DART_AWAIT(testAsyncProgramming());
      testGenerics();
      testExceptionHandling();
      testAdvancedFeatures();
      dart_print((dart_string("=") * dart_int(60)));
      dart_print(dart_string("✅ 所有测试完成"));
}
    {
      VMServiceEmbedderHooks::cleanup = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::createTempDir = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::ddsConnected = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::ddsDisconnected = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::deleteDir = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::writeFile = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::writeStreamFile = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::readFile = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::listFiles = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::serverInformation = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::webServerControl = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::acceptNewWebSocketConnections = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::serveObservatory = /* Constant: StaticTearOffConstant */;
      VMServiceEmbedderHooks::getResidentCompilerInfoFile = /* Constant: StaticTearOffConstant */;
      server = ObjectPtr<Server>(new Server(VMService::(), _ip, _port, _originCheckDisabled, _authCodesDisabled, _serviceInfoFilename, _enableServicePortFallback));
      if (_autoStart) {
      _toggleWebServer();
}
      _registerSignalHandler();
}
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
